### PR TITLE
Updated has_eulerian_path 

### LINKED ISSUE #4244: Issue with networkx.has_eulerian_path()
This issue adds to #3976 

There appears to be a problem with the `has_eulerian_path` function. It returns the wrong answer on this example:
```
test_graph = nx.DiGraph()
test_graph.add_edges_from([(0, 1), (1,2), (0,2)])
print(nx.has_eulerian_path(test_graph))
print(list(nx.eulerian_path(test_graph)))
```
Output:
```
True
[(0, 0), (0, 1), (1, 2)]
```

This directed graph does not have a Eulerian path, and, adding to #3976, the path returned by `eulerian_path` is not an actual Eulerian path, since it (1) does not include the edge `(0,2)` and (2) includes a self-loop `(0,0)` that is not in the graph.

My hunch is that the problem is in this line: https://github.com/networkx/networkx/blob/3351206a3ce5b3a39bb2fc451e93ef545b96c95b/networkx/algorithms/euler.py#L256

I think changing that `2` to be `semibalanced_ins+semibalanced_outs` may do the trick, since the definition is
>"...every _other_ vertex has equal in\_degree and out\_degree..." 

Going to see about fixing and adding this test. Appreciate any clues, ideas or tips.

### PR DESCRIPTION
**Pull request by cpurmessur**
Fixes #4244

As @rossbar said, the description does not match the implementation for a directed graph. This pull request fixes this issue.
The new implementation adequately checks if  
> "every other vertex has equal in_degree and out_degree".
It also addresses another issue about isolated points. The definition states
> "and all of its vertices with nonzero degree belong to a single connected component of the underlying undirected graph."

The previous implementation ignored vertices that had "nonzero degree". In this one, I created a copy of the graph and removed the isolated points before checking if it was weakly connected.

The code works on the test case that @tlarock brought up. However, this test case has not yet been added to test_euler.py

### CODE REVIEW

**Comment by cpurmessur** [comment]
I added the test case that @tlarock mentioned in issue #4244 in test_euler.py
I also added another test for isolated nodes.
Thanks, @dschult for your help with this!

File: networkx/algorithms/euler.py [hunk 1]
```diff
@@ -246,16 +246,25 @@ def has_eulerian_path(G):
     eulerian_path
     """
     if G.is_directed():
-        ins = G.in_degree
-        outs = G.out_degree
-        semibalanced_ins = sum(ins(v) - outs(v) == 1 for v in G)
-        semibalanced_outs = sum(outs(v) - ins(v) == 1 for v in G)
+        H = G.copy()
+        nodes_remove = [v for v in H if H.in_degree(v) == 0 and H.out_degree(v) == 0]
+        H.remove_nodes_from(nodes_remove)
```
Comment (dschult) [inline networkx/algorithms/euler.py:251]: We might be able to save a "copy" operation when there aren't any isolated nodes.
Something like:
```
        nodes_remove = [v for v in G if ins[v] == 0 and outs[v] == 0]
        if nodes_remove:
            G = G.copy()
            G.remove_nodes_from(nodes_remove)
```
and then shift references to H to be references to G.

Note that ```ins[v]``` is faster than ```ins(v)``` when there is only one node to look up.
Could be done a few lines down too... probably not a big deal...

**Review by dschult — APPROVED** [review]
This looks good -- thanks for the tests and the refactoring to remove the errors.

File: networkx/algorithms/euler.py [hunk 2]
```diff
@@ -246,16 +246,26 @@ def has_eulerian_path(G):
     eulerian_path
     """
     if G.is_directed():
+        nodes_remove = [v for v in G if G.in_degree[v] == 0 and G.out_degree[v] == 0]
```
Comment (rossbar) [inline networkx/algorithms/euler.py:249]: A comment here on why the if-statement might be worthwhile for future readers of the code, something like `# Remove isolated nodes (if any) without altering input graph`

**Review by rossbar — APPROVED** [review]
LGTM - thanks @cpurmessur !

One comment about potentially adding a comment to explain one of the "setup" steps, but feel free to take-it-or-leave-it

**Comment by cpurmessur** [comment]
No worries! Happy to help!

**Pull request merged**